src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1498 @@
+                return 0;
+
+        r = safe_atou32(p, &id);
+        if (r < 0 || id == 0)
+                return 0;
+
+        j = hashmap_get(m->jobs, UINT32_TO_PTR(id));
+        if (!j)
+                return 0;
+
+        *found = j;
+        return 1;
+}
+
+static int job_node_enumerator(
+                sd_bus *bus,
+                const char *path,
+                void *userdata,
+                char ***nodes,
+                sd_bus_error *error) {
+
+        _cleanup_strv_free_ char **l = NULL;
+        Manager *m = ASSERT_PTR(userdata);
+        Job *j;
+        unsigned k = 0;
+
+        l = new0(char*, hashmap_size(m->jobs) + 1);
+        if (!l)
+                return -ENOMEM;
+
+        HASHMAP_FOREACH(j, m->jobs) {
+                l[k] = strdup(j->object_path);
+                if (!l[k])
+                        return -ENOMEM;
+                k++;
+        }
+
+        *nodes = TAKE_PTR(l);
+        return 1;
+}
+
+static const sd_bus_vtable job_vtable[] = {
+        SD_BUS_VTABLE_START(0),
+
+        SD_BUS_PROPERTY("Id", "u", NULL, offsetof(Job, id), SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("Type", "u", job_property_get_type, offsetof(Job, type), SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("Offline", "b", NULL, offsetof(Job, offline), SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("Progress", "u", NULL, offsetof(Job, progress_percent), SD_BUS_VTABLE_PROPERTY_EMITS_CHANGE),
+
+        SD_BUS_METHOD("Cancel", NULL, NULL, job_method_cancel, SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_VTABLE_END,
+};
+
+static const BusObjectImplementation job_object = {
+        "/org/freedesktop/sysupdate1/job",
+        "org.freedesktop.sysupdate1.Job",
+        .fallback_vtables = BUS_FALLBACK_VTABLES({job_vtable, job_object_find}),
+        .node_enumerator = job_node_enumerator,
+};
+
+static Target *target_unref(Target *t) {
+        if (!t)
+                return NULL;
+
+        if (t->manager)
+                hashmap_remove(t->manager->targets, t->id);
+
+        free(t->name);
+        free(t->path);
+        free(t->version);
+        free(t->id);
+
+        return mfree(t);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(Target*, target_unref);
+
+static int target_new(TargetClass class, const char *name, Manager *m, Target **ret) {
+        _cleanup_(target_unrefp) Target *t = NULL;
+        int r;
+
+        assert(m);
+        assert(ret);
+
+        t = new(Target, 1);
+        if (!t)
+                return -ENOMEM;
+
+        *t = (Target) {
+                .manager = m,
+                .class = class,
+                .image_type = _IMAGE_TYPE_INVALID,
+        };
+
+        t->name = strdup(name);
+        if (!t->name)
+                return -ENOMEM;
+
+        if (class == TARGET_HOST) {
+                t->id = strdup("host");
+                if (!t->id)
+                        return -ENOMEM;
+        } else if (asprintf(&t->id, "%s/%s", target_class_to_string(class), name) < 0)
```

Comment:
use strjoin() for concatenating strings

Reply:
I'm not just concatenating here. There's a / in between

Reply:
strjoin(x, y, "z", w) works

Takes any number of args